en/ros/ros2_comm.md

```diff
@@ -1,64 +1,35 @@
-- `px4_msgs` contains PX4 ROS message definitions.
-  When this project is built it generates the corresponding ROS 2-compatible typesupport, used by ROS 2 nodes, and IDL files, used by `fastddsgen` to generate the microRTPS agent code.
-- `px4_ros_com` contains the microRTPS agent code templates for the agent publishers and subscribers.
-  The build process runs a `fastddsgen` instance to generate the code for the `micrortps_agent`, which compiles into a single executable.
+![Architecture XRCE-DDS with ROS 2](../../assets/middleware/xrce_dds/architecture_xrce-dds_ros2.svg)
 
-The PX4 Autopilot project automatically updates [`px4_msgs`](https://github.com/PX4/px4_msgs) with new message definitions whenever they are changed (in the `master` branch).
+<!-- doc source: https://docs.google.com/drawings/d/1WcJOU-EcVOZRPQwNzMEKJecShii2G4U3yhA3U6C4EhE/edit?usp=sharing -->
 
-:::note
-The subset of uORB topics that will be accessible to ROS applications can be found in the [bridge configuration yaml file](https://github.com/PX4/px4_ros_com/blob/main/templates/urtps_bridge_topics.yaml).
-:::
+The XRCE-DDS middleware consists of a client running on PX4 and an agent running on the companion computer, with bi-directional data exchange between them over a serial, UDP, TCP or custom link.
+The agent acts as a proxy for the client to publish and subscribe to topics in the global DDS data space.
 
-PX4 firmware contains the microRTPS client based on its build-time message definitions.
+The PX4 [microdds-client](../modules/modules_system.md#microdds-client) is generated at build time and included in PX4 firmare by default.
+It includes both the "generic" XRCE-DDS client code, and translation code that it uses to publish to/from uORB topics.
+The subset of uORB messages that are generated into the client are listed in [PX4-Autopilot/src/modules/microdds_client/dds_topics.yaml](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/microdds_client/dds_topics.yaml).
+The generator uses the uORB message definitions in the source tree: [PX4-Autopilot/msg](https://github.com/PX4/PX4-Autopilot/tree/main/msg) to create the message definitions.
 
-:::note
-Astute readers will note that the generated agent might not have been built with that same set of definitions (unless they were both built of the same 'master' commit).
 
-Right now this is not a problem because the PX4 message set/definitions are relatively stable and the updated/new messages get automatically deployed to `px4_msgs`.
+ROS 2 needs to have the same message definitions that were used to create the client.
+When working with PX4 main and normal releases you can clone the [PX4/px4_msgs](https://github.com/PX4/px4_msgs) package, which contains the definitions for each release in separate branches.
+If working on a custom build you could simply copy the msg definitoins directly out of PX4 source tree.
+These definitions can then be built in any ROS2 workspace that needs them.
```

Comment:
FMI I see in the examples that you just include these in the workspace and build:
```
cd ~/px4_ros_xrce_dds_ws/src
git clone https://github.com/PX4/px4_msgs.git
git clone https://github.com/eProsima/Micro-XRCE-DDS-Agent.git
source /opt/ros/humble/setup.bash
colcon build
```
1. How does the "system" know about these definitions? I.e. the agent is pre-built and the files are just msgs, so how does ROS2 know that it can use them for message definitions?
2. Do they have to be in a particular location to be built or would any files in a directory src be built?
3. What does it actually do with them - build IDE files then ... what?

Basically its clear that the agent needs to know about the messages and so does anything in ROS2 that wants to use them. What isn't clear is how this works. It looks like magic.

In previous version you could have a "generic DDS message" or "ROS2 message". My understanding is we now just have one format which matches ROS2. What does this mean for "generic DDS"? Is it possible to create a system that uses PX4 topics without ROS? Is it something we want to support?

Reply:
@hamishwillee 

> Basically its clear that the agent needs to know about the messages and so does anything in ROS2 that wants to use them.

The Agent does not need the message definitions at build time, nor at run time as they are necessary only to the ROS2 nodes (and to commands like `ros2 topic echo`) that want to subscribe and publish to them.
My way to see this is that the pair client-agent (and I'm talking about both micro XRCE-DDS Agent and microROS Agent) acts a unique ROS2 node physically distributed between PX4 and the companion computer. Therefore you need the message definitions only where you use the messages which is the client side. The Agents are just bridging the transport protocols.

> How does the "system" know about these definitions? I.e. the agent is pre-built and the files are just msgs, so how does ROS2 know that it can use them for message definitions?

On the client side https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/microdds_client/dds_topics.yaml not only specifies which uorb topics have to be exposed to ROS2, it also tells which is their message type:
```yaml
  - topic: /fmu/out/vehicle_odometry
    type: px4_msgs::msg::VehicleOdometry
```
Then ROS2 will know that it has to look for the message `px4_msgs::msg::VehicleOdometry` when it wants to interact with the topic `/fmu/out/vehicle_odometry`. The package `px4_msgs` contains a message definition named `VehicleOdometry.msg`which according to the ROS2 convention is compiled into the message named `px4_msgs::msg::VehicleOdometry`... so everything works out of the box!
You only need to take care that the files `VehicleOdometry.msg` in `px4_msgs/msg/` and the one in `PX$-Autopilot/msg/` are identical.

> Do they have to be in a particular location to be built or would any files in a directory src be built?

`colcon` build everything (unless you tell it otherwise) that resides in `src/`. Therefore as long are you have the package `px4_msgs` somewhere inside your workspace `src` folder you are fine, it will be compiled. However, it is not necessary to have it in the same workspace of the Agent: workspace can be linked together. You can create a workspace with just the message package and build it. Then you can move to another workspace and have the new one extending the previous with `source workspace_with_onl_px4_msgs/install/local_setup.bash`.

> What does it actually do with them - build IDE files then ... what?

This is going above my knowledge, when a `*.msg` file is compiled you are generating all the code needed to serialize/deserialize it, access it, the structure that contains all the message fields, ecc such that your application will be able to include the message header and create publishers and subscribers.


> My understanding is we now just have one format which matches ROS2

You are right!

> What does this mean for "generic DDS"?

Generic DDS means that your applications directly interface with the DDS network without using ROS, which is basically what the client on the PX4 side does. Everything works because ROS2 is using the exact same network.

> Is it possible to create a system that uses PX4 topics without ROS?

I guess it is, but I'm not 100% sure.

>  Is it something we want to support?

I don't know.

Reply:
@beniaminopozzan Thanks. So I get how the client side works - I can see how it knows about what messages to generate client code from the yaml , and it has the definitions.

However the agent doesn't see the yaml, all it sees is data. ROS2 does not need the msg to get the listing of topics, so I infer from this that it can determine message name/type from the data it receives. However it does need the definitions to echo a particular topic "detail"

The bit that is hard is this:

> You only need to take care that the files VehicleOdometry.msg in px4_msgs/msg/ ...

What I think you are saying is that as long as the messages are in a folder named "px4_msgs/msg" within a sourced workspace that ROS can see it will build them - right? 

Further that it gets this from the msg type: `px4_msgs::msg::`

And that the type comes with the data that ROS2 can interpret without the messages.

That makes sense!

Reply:
@hamishwillee of course!

> However the agent doesn't see the yaml, all it sees is data. ROS2 does not need the msg to get the listing of topics, so I infer from this that it can determine message name/type from the data it receives. However it does need the definitions to echo a particular topic "detail"

Exactly!

> What I think you are saying is that as long as the messages are in a folder named "px4_msgs/msg" within a sourced workspace that ROS can see it will build them - right?

You are not wrong, but what I meant is that the the message definitions used in ROS2 must be identical to the ones used in PX4. And indeed `PX4-bot` makes sure that this happens. But if a user decides to implement its own uorb message and to expose it to ROS2 then it has to

- update the yaml file
- copy the new message definition in "px4_msgs/msg" and recompile the workspace